File: DC_CS422_Lab2/Week2Lab-iOS/Week2LabTests/Week2LabTests.swift
//
//  Week2LabTests.swift
//  Week2LabTests
//
//  Created by Andrew Taylor on 1/15/23.
//

import XCTest
@testable import Week2Lab

final class Week2LabTests: XCTestCase {
    
    func testFlashcard() throws {
        let flashcards = getFlashcards()
        assert(flashcards.count == 10)
        assert(flashcards is Array<Flashcard>)
        
        let mirror = Mirror(reflecting: flashcards[0])
        assert(mirror.children.count == 2)
        assert(
            mirror.children.contains { (label: String?, _: Any) in
                label == "term"
            }
        )
        assert(
            mirror.children.contains { (label: String?, _: Any) in
                label == "definition"
            }
        )
    }

    func testFlashcardSet() throws {
        let flashcardSets = getFlashcardSets()
        assert(flashcardSets.count == 10)
        assert(flashcardSets is Array<FlashcardSet>)
        
        let mirror = Mirror(reflecting: flashcardSets[0])
        assert(mirror.children.count == 1)
        assert(
            mirror.children.contains { (label: String?, _: Any) in
                label == "title"
            }
        )
    }
}

File: DC_CS422_Lab2/Week2Lab-Android/app/src/test/java/cs/mad/week2lab/Week2LabTests.kt
package cs.mad.week2lab

import cs.mad.week2lab.entities.Flashcard
import cs.mad.week2lab.entities.FlashcardSet
import cs.mad.week2lab.entities.getFlashcardSets
import cs.mad.week2lab.entities.getFlashcards
import org.junit.Test
import kotlin.reflect.KProperty

class Week2LabTests {

    @Test
    fun testFlashcard() {
        assert(sizeAndType<Flashcard>(getFlashcards()))

        assert(Flashcard::term is KProperty<String>)
        assert(Flashcard::definition is KProperty<String>)
    }

    @Test
    fun testFlashcardSet() {
        assert(sizeAndType<FlashcardSet>(getFlashcardSets()))
        assert(FlashcardSet::title is KProperty<String>)
    }

    private fun <T> sizeAndType(array: Array<T>) = sizeAndType(array.asList())
    private fun <T> sizeAndType(collection: Collection<T>) = collection.size == 10
    private fun <T> sizeAndType(other: Any) = false
}